[tensorexpr] Fix the buf size of discontiguous tensors

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69657

Differential Revision: [D32974473](https://our.internmc.facebook.com/intern/diff/D32974473)

huiguoo committed:
[tensorexpr] Fix the buf size of discontiguous tensors

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32974473).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -937,9 +937,13 @@ Tensor TensorExprKernel::bindInput(const torch::jit::Value* input) {
       BufHandle inBuffer(
           "t" + input_name_map_[input],
-          {0},
+          {flat_size},
```
Nit: might be worth running IRSimplifier on `flat_size`.

ZolotukhinM reviewed (approved):
Awesome!

davidberard98 commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -937,9 +937,13 @@ Tensor TensorExprKernel::bindInput(const torch::jit::Value* input) {
       }
+      ExprHandle flat_size = 0;
+      for (size_t i = 0; i < *tt->sizes().size(); i++) {
+        flat_size = flat_size + *tt->sizes()[i] * *tt->strides()[i];
```
should this actually be `max(flat_size, sizes[i] * strides[i])` ? e.g. a 2x4 tensor with strides [4, 1] has size 8 (2\*4), not 12 (2\*4 + 4\*1)

ZolotukhinM replied:
I think technically it should be `sum((sizes[i] - 1)*strides[i]) + 1` - on every dimension we can attempt to access element `sizes[i]-1` and as we're accessing it we use `strides[i]`. In [2,4]-[4,1] example this would yield `1*4 + 3*1 + 1 = 8`.

huiguoo replied:
@ZolotukhinM is right about this! Updated.

huiguoo committed:
Update on "[tensorexpr] Fix the buf size of discontiguous tensors"


Differential Revision: [D32974473](https://our.internmc.facebook.com/intern/diff/D32974473)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32974473).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -937,9 +937,14 @@ Tensor TensorExprKernel::bindInput(const torch::jit::Value* input) {
       }
+      ExprHandle flat_size = 1;
+      for (size_t i = 0; i < *tt->sizes().size(); i++) {
+        flat_size = flat_size + (*tt->sizes()[i] - 1) * *tt->strides()[i];
```
Now since we have -1 there we might need to check if `size[i]` is not 0.

huiguoo replied:
That's interesting. If one dimension has size 0, then the flat size of the tensor should be 0 right?

huiguoo replied:
> the corresopnding IR will be simplified at some later point anyway

I guess we will treat `inBuffer.load(idx)` as dead code and remove it in simplifier? This makes sense.

ZolotukhinM replied:
Yes and yes :)

huiguoo committed:
Update on "[tensorexpr] Fix the buf size of discontiguous tensors"


Differential Revision: [D32974473](https://our.internmc.facebook.com/intern/diff/D32974473)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32974473).

eellison commented:
What is the context of this ? Could you add something to commit message
